### DISCUSSION TITLE
Introducing UPDI4AVR-USB

### DISCUSSION THREAD

**Discussion by askn37** [avrdudes/avrdude#1878] [Show and tell]
Introducing a new firmware that easily transforms AVR-DU family into UPDI/TPI writers with USB 2.0 full speed connection. It uses JTAGICE3 command set with EDBG/CMSIS-DAP protocol so it can be used like PICKit4, Curiosity, Xplained.

First commit is now live. The pre-built binaries can be uploaded to the ["AVR64DU32 Curiosity Nano : EV59F82A"](https://www.microchip.com/en-us/development-tool/ev59f82a) product for easy setup. (Other model numbers require rebuilding.)

- https://github.com/askn37/UPDI4AVR-USB/blob/main/README_en.md

<img src="https://askn37.github.io/product/UPDI4AVR/images/IMG_3832.jpg" width="400"> <img src="https://askn37.github.io/product/UPDI4AVR/images/U4AU_UPDI.drawio.svg" width="400">

AVRDUDE>=7.3 is required to configure AVR-DU families, but once configured, AVRDUDE>=7.2 can also be used as long as you are targeting non-AVR-Ex/AVR-DU families. AVRDUDE>=main is required for AVR-Ex families.

Oh! Don't forget to write the VID:PID to the EEPROM that matches the programmer you want to emulate!
(e.g. `pickit4==03EB:2177`)
If AVRDUDE>=main you can use `-P usb:04DB:0B15 -c pickit4_updi`, but that wasn't the case before.

VCP can also be used simultaneously. Total bandwidth is 500kbps to 1Mbps. DTR linked reset function is also included, and Arduino IDE bootloader access is also possible.

<details>

<img width="800" alt="SS" src="https://github.com/user-attachments/assets/ceed457c-853e-4127-89d2-327a36e47207">

</details>

**Comment by MCUdude** [comment]
Impressive work! Very very cool! I'll definitely try this on my AVR-DU Curiosity Nano!

BTW how do you deal with HV reset? The AVR-DU obviously needs to apply a high voltage (usually 12V) to either the UPDI pin or the reset pin. Do you have a reference schematic for how this is done?

**Reply by askn37** [reply]
> Impressive work! Very very cool! I'll definitely try this on my AVR-DU Curiosity Nano!
> 
> BTW how do you deal with HV reset? The AVR-DU obviously needs to apply a high voltage (usually 12V) to either the UPDI pin or the reset pin. Do you have a reference schematic for how this is done?

The HV control technology of UPDI12V and RST9V has already been realized in previous research.

- [UPDI4AVR](https://askn37.github.io/product/UPDI4AVR/README_en.html)
- [Schematic diagram](https://askn37.github.io/product/UPDI4AVR/2306_UPDI4AVR/2306_Zinnia-UPDI4AVR-MZU2406B2.pdf) (PDF)

This design works fine up to the AVR-EA family, but be aware that the AVR-EB and AVR-DU families may slightly exceed the absolute ratings.

Why? Read the respective product datasheets carefully. The AVR-EB family has a lower *absolute rating for the RESET pin of 9V*. The __AVR-DU family__ goes further and for the first time specifies a typical voltage of __VCC+2V__, where previously the typical voltage was not stated anywhere.

For this reason, the conventional method of clipping the maximum voltage with a Zener diode has become difficult to support all AVR-Dx/Ex families. Active variable voltage output is required, and the circuit needs to be redesigned. In addition, the number of control terminals is insufficient, so some of them must be replaced with inverters. This is my future challenge.

Currently AVRDUDE has no mechanism to inform the programmer of the high voltage setting for AVR models it can't even read the SIB for, so you have to provide at least the second and third bytes of the signature of the model you're working with so the programmer can set the correct voltage. These were completely absent in traditional JTAGICE, so you have to extend them yourself (except for PICKit5, which controls it in a completely different way).

**Reply by askn37** [reply]
Supplementary note.

When the AVR-Dx family was replaced by the "full" datasheet from the "preliminary" datasheet, it was clarified that the RESET pad is rated at 8.5V absolute and 7.5V typical, which was unclear before. It was previously written as 9V. This specification should also apply to AVR-Ex. (So the Zener diode had to be changed to 8.2V or less.)

On the other hand, the tinyAVR-1 and earlier states that UPDI is 11.5V to 12.5V, but the tinyAVR-2 family is unclear due to the lack of this description. Experimentally, it is possible to enter HV control at a lower voltage.

**Reply by MCUdude** [reply]
Thank you for your detailed explanation! I wonder how the PICkit4/5's do this! UPDI HV programming seems like a minefield to navigate through. One theoretical option would be to have a 12V source available and use a programmable zener of some sort (TL431?) that the AVR-DU could drive.

Again, I'm very impressed with the UPDI4AVR project, and I'm sure it will become popular. I know English isn't your primary language (not mine either, mine is Norwegian), but I think some English documentation at the top of the [UPDI4AVR repo](https://github.com/askn37/UPDI4AVR-USB) could help the project get more users, and probably be picked up by sites like Hackaday. The JTAG2UPDI project quickly became very popular because of the documentation, and how easy it was to build, program, and use. I think this project _could_ achieve something close to that 🙂

I'm planning to design a dedicated UPDI4AVR programmer to use for testing. Probably something similar to the [SerialUPDI programmer](https://www.tindie.com/products/mcudude/serialupdi-programmer/) in terms of looks and features.

**Comment by askn37** [comment]
Subsequent update:
I have completed the prototype of the controller board (and its case) for UPDI and TPI HV control, which is very nice and works without any issues. Also, ATtiny20/40 has been added to the verification device list. These have recently become NRD, so I don't think many people will adopt them, but it's good to be able to use them.

<img src="https://github.com/user-attachments/assets/3405cbe1-ba6d-4742-b302-2f0c23920115" width=400>

Control of the PDI interface is still in progress. Unfortunately, due to electrical constraints, it cannot be controlled via the HV control connector (open drain type) and requires a dedicated wired connector. Furthermore, there is no way to permanently change the CNANO to run at 3.3V, so it's not something anyone can easily try.

<img src="https://github.com/user-attachments/assets/225333cc-c48a-493a-9199-b7fb2d0ef70e" width=400>

Once these difficulties were overcome, it was possible to achieve full PDI control of at least the ATxmega128A4U.

<details>

```
./avrdude_preq/tools/test8 -e./avrdude_preq/build_darwin/src/avrdude -a-Pusb:04DB:0B15 -r1  pickit4_pdi x128a4u
Reading 141643 bytes for multiple memories from input file bak-pickit4_pdi-x128a4u.hex
Writing 2048 bytes to eeprom
Writing | ################################################## | 100% 0.82 s 
Reading | ################################################## | 100% 0.26 s 
2048 bytes of eeprom verified
Writing 139264 bytes to flash
Writing | ################################################## | 100% 16.39 s 
Reading | ################################################## | 100% 8.71 s 
139264 bytes of flash verified
Writing 131072 bytes to application
Writing | ################################################## | 100% 15.39 s 
 | ################################################## | 100% 8.23 s 
131072 bytes of application verified
Writing 4096 bytes to apptable
Writing | ################################################## | 100% 0.48 s 
 | ################################################## | 100% 0.26 s 
4096 bytes of apptable verified
Writing 8192 bytes to boot
Writing | ################################################## | 100% 0.96 s 
 | ################################################## | 100% 0.51 s 
8192 bytes of boot verified
Writing 7 bytes to fuses, 7 bytes written
Warning: fuses verification mismatch
  device 0x00 != input 0xff at addr 0x0006 (error)
Error: fuses verification mismatch
Writing 1 byte (0xFF) to fuse0, 1 byte written, 1 verified
Writing 1 byte (0x00) to fuse1, 1 byte written, 1 verified
Writing 1 byte (0xFF) to fuse2, 1 byte written, 1 verified
Writing 1 byte (0xFE) to fuse4, 1 byte written, 1 verified
Writing 1 byte (0xFF) to fuse5, 1 byte written, 1 verified
Writing 1 byte (0xFF) to lock, 1 byte written, 1 verified
Writing 256 bytes to usersig
Writing | ################################################## | 100% 0.03 s 
 | ################################################## | 100% 0.02 s 
256 bytes of usersig verified

Avrdude done.  Thank you.
test8: x128a4u initialised
```

```
./avrdude_preq/tools/test8 -e./avrdude_preq/build_darwin/src/avrdude -a-Pusb:04DB:0B15 pickit4_pdi x128a4u

Processing -U ALL:r:bak-pickit4_pdi-x128a4u.hex:I
Reading multiple memories ...
 - eeprom         | ################################################## | 100% 0.26 s 
 - flash          | ################################################## | 100% 8.72 s 
 - application    | ################################################## | 100% 8.28 s 
 - apptable       | ################################################## | 100% 0.26 s 
 - boot           | ################################################## | 100% 0.51 s 
 - fuses          | ################################################## | 100% 0.03 s 
 - fuse0          | ################################################## | 100% 0.00 s 
 - fuse1          | ################################################## | 100% 0.00 s 
 - fuse2          | ################################################## | 100% 0.00 s 
 - fuse4          | ################################################## | 100% 0.00 s 
 - fuse5          | ################################################## | 100% 0.00 s 
 - lock           | ################################################## | 100% 0.00 s 
 - prodsig/sigrow | ################################################## | 100% 0.01 s 
 - calibration    | ################################################## | 100% 0.02 s 
 - sernum         | ################################################## | 100% 0.06 s 
 - tempsense      | ################################################## | 100% 0.01 s 
 - usersig        | ################################################## | 100% 0.02 s 
 - signature      | ################################################## | 100% 0.00 s 
Writing 285028 bytes to output file bak-pickit4_pdi-x128a4u.hex

Processing -T fact reset
performing chip erase; discarded pending writes to flash and EEPROM
Caching | ################################################## | 100% 0.26 s 
Caching | ################################################## | 100% 0.01 s 
Caching | ################################################## | 100% 0.02 s 
Synching cache to device ... done
after the next reset the part will have F_CPU = 2.000 MHz

Processing -U ALL:w:bak-pickit4_pdi-x128a4u.hex:I
Reading 141643 bytes for multiple memories from input file bak-pickit4_pdi-x128a4u.hex
Writing 2048 bytes to eeprom
Writing | ################################################## | 100% 0.83 s 
Reading | ################################################## | 100% 0.26 s 
2048 bytes of eeprom verified
Writing 139264 bytes to flash
Writing | ################################################## | 100% 14.14 s 
Reading | ################################################## | 100% 8.76 s 
139264 bytes of flash verified
Writing 131072 bytes to application
Writing | ################################################## | 100% 13.41 s 
0000 | ################################################## | 100% 8.22 s 
131072 bytes of application verified
Writing 4096 bytes to apptable
Writing | ################################################## | 100% 0.42 s 
0000 | ################################################## | 100% 0.26 s 
4096 bytes of apptable verified
Writing 8192 bytes to boot
Writing | ################################################## | 100% 0.83 s 
0000 | ################################################## | 100% 0.51 s 
8192 bytes of boot verified
Writing 7 bytes to fuses, 7 bytes written, 7 verified
Writing 1 byte (0xFF) to fuse0, 1 byte written, 1 verified
Writing 1 byte (0x00) to fuse1, 1 byte written, 1 verified
Writing 1 byte (0xFF) to fuse2, 1 byte written, 1 verified
Writing 1 byte (0xFE) to fuse4, 1 byte written, 1 verified
Writing 1 byte (0xFF) to fuse5, 1 byte written, 1 verified
Writing 1 byte (0xFF) to lock, 1 byte written, 1 verified
Writing 256 bytes to usersig
Writing | ################################################## | 100% 0.02 s 
0000 | ################################################## | 100% 0.02 s 
256 bytes of usersig verified

Processing -U ALL:v:bak-pickit4_pdi-x128a4u.hex:I
Verifying 141643 bytes of multiple memories against input file bak-pickit4_pdi-x128a4u.hex
Reading | ################################################## | 100% 0.26 s 
2048 bytes of eeprom verified
Reading | ################################################## | 100% 8.72 s 
139264 bytes of flash verified
Reading | ################################################## | 100% 8.19 s 
131072 bytes of application verified
Reading | ################################################## | 100% 0.26 s 
4096 bytes of apptable verified
Reading | ################################################## | 100% 0.51 s 
8192 bytes of boot verified
7 bytes of fuses verified
1 byte of fuse0 verified
1 byte of fuse1 verified
1 byte of fuse2 verified
1 byte of fuse4 verified
1 byte of fuse5 verified
1 byte of lock verified
Reading | ################################################## | 100% 0.01 s 
64 bytes of prodsig/sigrow verified
4 bytes of calibration verified
14 bytes of sernum verified
2 bytes of tempsense verified
Reading | ################################################## | 100% 0.02 s 
256 bytes of usersig verified
3 bytes of signature verified

Avrdude done.  Thank you.
```

</details>

The test8 script log seems to be partially distorted by the PDI, the read and verify labels are weird, the fuse size seems to be set to 7 when it should be 6, and it's trying to write to the ROM area.

**Reply by askn37** [reply]
P.S.

CNANO confirmed that updating the firmware to at least `1.31 (rel. 39)` or higher will permanently remember `-x vtarg=<dbl>` and preserve the VTG voltage even after powering off. This is very important for PDI support. Don't forget to cast the following incantation before actually wiring the Xmega device, so that the target won't be destroyed.

```
avrdude -cpkobn_updi -v -pavr64du32 -xvtarg=3.3
```

*It seems that there are only three valid choices: 5.0, 3.3, and 1.8. UPDI devices can write to memory at anything other than 1.8, and TPI devices can only write memory correctly at 5.0. In this respect, PDI and TPI are mutually exclusive.*

**Reply by stefanrueger** [reply]
@askn37 Love the design! Looks vvv neat!

> the fuse size seems to be set to 7 when it should be 6

That could be changed in `avrdude.conf`. The comment here https://github.com/avrdudes/avrdude/blob/0b92721af9aaa686095542cc54715a3fa06b5895/src/avrdude.conf.in#L15763-L15773 supports your observation. Would need checking data sheets first.

**Comment by askn37** [comment]
Latest Updates:

Support has been added for AVR-Lx (NVMCTRL version 5) and AVR-Sx (NVMCTRL version 6). Operation has also been verified for the AVR32LA32 and AVR32SD32. Consequently, all devices currently listed in ATpack—including the Sx series mentioned on the download page—are now supported.

Other:

I have created an experimental bootloader for the AVR-SD. The presence of ECC memory protection presents a challenge; as the data range targeted for rewriting expands, the time required to recalculate parity increases, apparently causing frequent CPU clock stalls. This behavior contrasts with UPDI-based rewriting, where recalculation is (presumably) deferred until a system reset.

**Comment by askn37** [comment]
Recent developments: (Aug. 2026)

I have added limited ISP control functionality, as the need arose for my own use; however, only a very small number of chips have been verified. The standard ICSP-6P wiring fully overlaps with the UPDI/TPI pinout. Additionally, extra wiring allows for supplying an external clock to the target device, enabling fuse recovery unless the device is severely damaged. Test results are available in the discussion below.

https://github.com/askn37/UPDI4AVR-USB/discussions/18

![](https://camo.githubusercontent.com/fa1f4a84ff12077c52f926c9ca3648ef5c03a870547e3df2fcf561a28288602a/68747470733a2f2f61736b6e33372e6769746875622e696f2f70726f647563742f55504449344156522f696d616765732f553441555f50494e4f55542e64726177696f2e737667)

**Reply by stefanrueger** [reply]
Very nice. Thanks for keeping us informed about progress in that amazing project.
